docs/rfc/rfc-101-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,436 @@
+
+```golang
+
+func (mem* ClistMempool) banPeer(peerID NodeID) {
+    numFails := 0
+      // Check whether this peer has sent us transactions that fail
+    if val, ok := mem.peerFailureMap[peerID]; ok {
+      lastFailureT := val.lastFailure
+      numFails = val.numFails
+    // if the failure was recent enough, update the number of failures and 
+    // ban peer if applicable
+    if time.Since(lastFailureT) <= failureResetInterval {
+      if numFails == maxAllowedFailures - 1 {
+          // Send Ban request to p2p 
+        }
+      }
+    }
+    // Update the time of the last failure     
+    mem.peerFailureMap[peerID] = { time.Now(), numFailures + 1}
+}
+```
+If transactions with `ResponseCheckTx.code > 1` are not deleted from the cache and we want to ban them on the first offence,
+we can skip the second `if` in the code above but call `banPeer` immediately at the end of the function. The function `banPeer` should simply forward the `peerID` to the p2p layer. In fact, we do not need to store any information on this peer as the node will remove it from its peer set.
+
+**Signaling the ban to p2p**
+
+As it is the mempool **reactor** that has access to the p2p layer, not the actual mempool implementation, the peer banning function will most likely have to send the peer ID to a channel to inform the reactor of the fact that this peer should be banned. This would require adding a channel for communication into the `CListMempool` on construction, and adding a routine in the mempool reactor that waits on a response via this channel. However, the actual implementation of this is yet to be defined.
+
+**Implementing peer banning on recheck**
+
+Currently the recheck logic confirmes whether once **valid** transactions, 
+, where `ResponseCheckTx.code == 0`, are still valid. 
+
+As this logic loops through the transactions in any case, we can leverage it to check whether we can ban peers. 
+However, this approach has several downsides:
+- It is not timely enough. Recheck is executed after a block is committed, leaving room for a bad peer to send 
+us transactions the entire time between two blocks. 
+- If we want to keep track of when peers sent us a traansaction and punish them only if the misbehaviour happens 
+frequently enough, this approach makes it hard to keep track of when exactly was a transaction submitted. 
+- Rechecking if optional and node operators can disable it. 
+- Furthermore, rechecking is a node local configuration parameter. This means that, some nodes might be performing this check,
+  while others will be unaware of this. 
+
+On the plus side this would avoid adding new logic to the mempool caching mechanism and keeping additional information about
+transaction validity. But we would still have to keep the information on peers and the frequency at which they send us bad transactions.
+
+Transactions that became invalid on recheck should not be cause for peer banning as they have not been gossiped as invalid transactions. 
```

Comment:
This is not necessarily true. With prepare and process proposal, bad actors could send transactions that are valid but not ripe for proposal, which would always become invalid later. I think only the app can be the judge of it.